### DISCUSSION TITLE
Thoughts on providing a way for the adapter to support running debuggees under a user provided emulator

### DISCUSSION THREAD

**Discussion by calewis** [microsoft/debugpy#1684] [Enhancements]
Would there be any interest in having debugpy support running debuggees under some other program?  My main goal would be to insert an emulator between the adapter and the debuggee when we can detect that the debuggee is not configured for the native architecture.  

My specific use case targets elf binaries that are being run via the adapter, making this detection fairly easy.   The way I have cobbled together a working pipeline is to replace the debuggee with /path/to/emulator in the command that will launch the debuggee's process.    

I'm mainly checking here to see if there is any interest in supporting the ability to place a shim program between the adapter and the debuggee in the upstream repository.

**Comment by rchiodo** [comment]
So the debuggees in this case aren't python code?

**Reply by calewis** [reply]
They are python code wrapped with a C++ launcher binary.

**Reply by calewis** [reply]
The closest looking thing to what I did in the upstream project would be something like the following here: https://github.com/microsoft/debugpy/blob/25955a05d87a6e3d7694f7479c1c17ded0f9af90/src/debugpy/launcher/handlers.py#L38

```
    python = request("python", json.array(str, size=(1,)))
    if user_provided_shim:
      cmdline = [user_provided_shim]
    cmdline += list(python)
```
instead of just command line.  In my version I'm currently using `elftools.elf.elffile` to detect if the debuggee architecture matches with the architecture of the adapter.  If not, we insert the emulator as the first argument. 

I'm not suggesting that upstream should adopt this feature, more just asking if there is interest in having it, or having a more general ability to insert a shim before python?

**Reply by rchiodo** [reply]
Yeah seems valuable to me. Something that would allow the request to start call something else first. 

I believe this is already possible. It's a configuration option. In that code there, it's what is returned by the 'python' value of the request.

In a vscode launch.json, it's set like so:

```json
    {
      "name": "Python Debugger: Current File",
      "type": "debugpy",
      "request": "launch",
      "program": "${file}",
      "python": "my executable here",
      "pythonArgs": ["args to exe1", "arg to exe 2"]
    }
```
